### PR TITLE
Urn history improvements

### PR DESCRIPTION
**Pull request by EdNoepel**
Implemented improvements to querying urn history:
- Maintain in-memory cache of urn history so it isn't continuously re-requested
- Use vulcanize full node to retrieve urn state rather than building it from urn operations
- When configured, only use vulcanize to initialize the cache

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -297,7 +296,8 @@ the queue to complete.  [Etherscan.io](etherscan.io) can be used to view your ac
 This keeper connects to the Ethereum network using [Web3.py](https://github.com/ethereum/web3.py) and interacts with
 the Dai Stablecoin System (DSS) using [pymaker](https://github.com/makerdao/pymaker).  A connection to an Ethereum node
 (`--rpc-host`) is required.  [Parity](https://www.parity.io/ethereum/) and [Geth](https://geth.ethereum.org/) nodes are
-supported over HTTP. Websocket endpoints are not supported by `pymaker`.
+supported over HTTP. Websocket endpoints are not supported by `pymaker`.  A _full_ or _archive_ node is required;
```
Comment (KentonPrescott) [inline README.md:299]: Thank you for clarifying which type of node is supported

File: README.md [hunk 2]
```diff
@@ -250,13 +250,12 @@ directly deposited to your token balance.
...
-    the `vat` contract to bulid a list of urns, and then check the status of each urn.  Setting this too low will
-    overburden your node.
- * **Deploy a [VulcanizeDB lite instance](https://github.com/makerdao/vdb-lite-mcd-transformers) to maintain your own
-    copy of urn state in PostgresQL, and then set `--vulcanize-endpoint` to your instance**.  This will conserve
+To start `flip` auctions, the keeper needs a list of urns and the collateralization ratio of each urn.  This list is 
+built by initializing an in-memory cache with urn history, and then querying recent events upon each block.  There are 
+two ways to initialize the cache:
+ * **Set `--from-block` to the block where the first urn was created** to scrape the chain for `frob` events. 
+ * **Deploy a [VulcanizeDB instance](https://github.com/makerdao/vdb-mcd-transformers) to maintain your own
+    copy of urn state** in PostgresQL, and then set `--vulcanize-endpoint` to your instance.  This will conserve
```
Comment (KentonPrescott) [inline README.md:258]: Although it's optional, it may be worth adding a comment on the new `--vulcanize-key` arg.

Comment (EdNoepel) [reply README.md:258]: Done.

**Comment by EdNoepel** [comment]
> Great job on the in-memory cache. Our of curiosity, do you have any performance numbers from this cache addition?
> 
> Adding a local cache (example that comes to mind is `tinydb`) could a be a potential next improvement, as it relieves the need to re-request after the keeper is terminated.
> Note that this would only be valuable for operators without access to `VulcanizeDB`.

Previously, without Vulcanize, the keeper would continuously request all history from the node.  Eventually, this became too expensive for ETH-A and BAT-A, forcing biters to run their own Vulcanize instance or write their own implementation to maintain urn state.  I updated pymaker's `past_frobs` to allow us to query in smaller chunks, which cost about 28 minutes to compile a list of urns, plus 8 minutes to get the state of each.  Requesting the last 10 blocks of frobs costs under 1 second with this setup.  These numbers are using a remotely-hosted Parity archive node.  To attach some adjectives, performance went from "unusable" to "sufficient". 

I have no plans to persist urn state to disk.  I was somewhat uncomfortable keeping it in memory, as I prefer the keeper to remain as stateless as possible.  Changes to Vulcanize forced this implementation.

**Pull request merged**